### ISSUE TITLE
Implement more gambits

### ISSUE THREAD

**Issue by tobscure** [flarum/core#256]
Gambits are little pieces of search criteria that can be specified in the search string. Just like GitHub's [search syntax](https://help.github.com/articles/search-syntax/).

Right now we've only implemented a few: `is:unread` and `author:username` in core, and `tag:foo,bar`, `is:followed`, `is:ignored`, `is:sticky`, and `is:locked` in their respective extensions.

We probably want to implement the following gambits:

* Multiple author search, separated by commas (`author:username1,username2`)
* `created:date` where `date` is the same syntax as on GitHub
* `updated:date`
* `contributor:username1,username2`
* `replies:number` where `number` is the same syntax as on GitHub

Pending #XX, we will also want to have an `in:title` gambit which scopes fulltext search to only look in discussion titles.

This is a precursor to an extensible advanced search interface which will come later.

**Comment by Albert221** [comment]
See #703.

**Comment by BartVB** [comment]
Can I put in a request for searching by IP address? We use that to track down spammers and their posts. Obviously this should be behind some kind of permission.

**Comment by franzliedke** [comment]
@BartVB Good idea. If @tobscure agrees, we can add that to the list at the top of this issue.

**Comment by luceos** [comment]
Contributor or participant is related to the author of a post, contrary to the current Author Gambit for Discussions.

**Comment by askvortsov1** [comment]
@jordanjay29 I'm assuming that the replies gambit is just for sorting, but if it's for filtering, I would propose splitting it into replies_gte and replies_lte (although not sure what the use case for that would be).

**Comment by askvortsov1** [comment]
This post was made a while ago, and hasn't been updated recently. The following have already been updated:

- Multiple author search
- Multiple contributor search (PR)
- Time Created
- Time Last Updated (PR)
- Improvements to fulltext title searching

There are a few more things to consider:
- Should the created and uploaded gambits be updated offer better support for "before" and "after"? eg: `created:before:2020-01-04`
- Do we even want a replies # gambit? This feels like something that would definitely need to be greater than or less than, I can't imagine searching for posts with exactly n replies
- Do we want to consider permissions in gambits? While it would be nice to search by IP address, that would mean hardwiring in a big chunk of new logic which, frankly, feels like overkill.

**Comment by luceos** [comment]
Looking at your discussion on internals (next time please stick to the issue, so that we are not scattering our path of thought everywhere), I think (in my opinion) we need to decide on one date; which to me would make most sense for created_at. I don't think anyone would ever have any need for the other fields and if they do they can either:

- use the database directly
- create their own gambit

Having said that, I like what discord did and the argument @datitisev made; keep it stupid simple. Increasing nesting of the namespace (semicolon) will not do our app any good. For this reason I think the PR #2027 should be closed.

PS. as a general remark, if you foresee these kind of decisions it's often better to first discuss the options and then do a PR. Otherwise you spent valuable time on something that might be discarded while risking feeling burned/overruled by a core team decision. Having said so, your efforts are immensely valued, keep it up!

**Comment by askvortsov1** [comment]
Sounds good! Just to clarify, are you proposing splitting created into multiple gambits to support created_before and created_after? I agree that nesting is kinda gnarly, but I'm also not sure how I feel about having 6 different date gambits. 

PR #2027 doesn't have nesting, and I'm not sure that "Updated" isn't useful: It can be good for seeing what conversations / users / areas of discussion are active / inactive with more specificity than just sorting it.

If y'all think we should close it, I'm perfectly fine with that: it was a good opportunity to get to know a part of the codebase I'd never seen before.
